README.org
#+title: Thermophysical and Thermochemical calculations

* Packages in this repository
  - Thermophysical and Thermochemical coefficients for gases and liquids
  - Collision integrals
  - Transport coefficients for low pressure gases

  Package Thermophysical and Thermochemical coefficients for gases and
  liquids contains data for:
  - Density and specific volume
  - C_p and C_v
  - Enthalpy and entropy
  - Internal energy and speed of sound
  - Viscosity and thermal conductivity
  - Joule-Thomson coefficient Surface tension (saturation curve only)

  This package does not include ab-initio methods for calculating the
  properties.  It only includes methods based on simple formulas or
  table interpolation.

  The data is obtained from on-line databases (such as the NIST
  webbook) and journal articles.

  All internal calculations and all results are a mixture of SI and
  chemistry friendly units - atomic mass units, etc.

  Package for collision integrals
  - Reads table of Lennard-Jones potential parameters
  - Provides fitting formulas for collision integrals Omega-11 and
    Omega-22

  This package is used by the transport coefficients.  As it is closer
  to gas kinetic theory, it uses SI units (mass is in kg, cross-sections
  in meters)  

* Warning
  Thermal conductivity calculated using Lennard-Jones collision
  integrals and Cv gives incorrect values for N2 by a factor of 1.25.
  The code gives correct values for Ar.

  I have tested the values of Cv for both Ar and N2 (which are
  calculated by the package).  I can't figure out where else the
  difference could originate form.

  Thermal conductivity calculation using the Lemmon-Jacobsen
  coefficients gives correct values of conductivity.
  
* Status

  Fairly stable, but lightly tested after recent refactoring of
  collision integrals.

  New features added as needed.

** Future plans



ChangeLog
2012-08-24  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/cck-transport-coefficients.lisp
	(Delta-1): Fixed docstring error

2012-07-31  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/transport-coefficients-user-package-def.lisp
	(#:transport-coefficients-user): Added package documentation

2012-07-27  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/generic-functions.lisp (lambda-ig2):
	Bug-fix.  Generic function had extra model parameter
	(mu2, lambda-ig2): Fixed documentation error

	* transport-coefficients/transport-coefficients-user-package-def.lisp
	(#:transport-coefficients-user): Importing symbols from
	gas-kinetics-physics-constants

	* transport-coefficients/transport-coefficients-user.asd
	(#:transport-coefficients-user): Added dependence on
	gas-kinetics-physics-constants package.

2012-07-18  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/generic-functions.lisp: Adapting
	following argument ordering: 1) species, 2) densities/pressures,
	3) temperature

	* transport-coefficients/mcc-diffusivity.lisp (delta-1): New
	method for 'mcc model.  This complements delta calculated for the
	cck model.

	* transport-coefficients/cck-transport-coefficients.lisp (D12-1):
	Factored out the numerical coefficient ``coeff'' into a let-form
	outside the two 'cck methods for hard-sphere and Lennard-Jones
	potentials.
	(D12-1): Bug-fix: Commented out an erroneus 1e3 coefficient that
	multiplied the mass.  That coefficient is absorbed in ``coeff''.
	This may explain the fudge factor introduced on 2012-07-13 in
	D12-mcc (see entry below)

2012-07-17  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/mcc-diffusivity.lisp (Delta-mcc): New
	function.
	(D12-mcc): Rewrote body to use Delta-mcc

2012-07-13  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/mcc-diffusivity.lisp (D12-mcc): I added a
	multiplier of sqrt 1e3 in order to get agreement with first order
	Chapan-Enskog.  I have not been able to trace the source of the
	discrepancy.  See my notes in the low-pressure-flow.org file.

	* transport-coefficients/cck-transport-coefficients.lisp (D12-1):
	For the cck model I added a method for the hard-sphere potential

2012-07-12  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* src/thermo-data.asd (:thermo-data): Removed collision integrals
	and lennard-jones references.  These are now part of a separate
	package

	* src/environment-setup.lisp (*data-directory*)
	(*janaf-directory*): Fixed selector to new windows workstation


2012-06-13  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/generic-functions.lisp (mu2 mu2-1): Updated
	documentation for the interface and generic function.

	* transport-coefficients/mcc-viscosity.lisp (mu2-1): Redefined
	returned values for the McCormack model.  Instead of returning mu1
	& mu2, I return the total viscosity.  Individual viscosities are
	avaiable via multiple values

2012-05-27  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/cck-transport-coefficients.lisp
	(alpha-T-1): Changed interface for cck to accomodate McCormack
	formula that requires both p1 and p2

	* transport-coefficients/generic-functions.lisp (D12b): Added
	function for binary diffusion for a gas mix, explicitly dependent
	on the potentials of each species, and their densities.  This
	function is necessary for higher order binary diffusion
	coefficients

2012-05-18  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* transport-coefficients/generic-functions.lisp (lambda-ig1)
	(lambda-ig2, lambda1): Refactoring of thermal conductivity.  I now
	have separated lambda for ideal gas from that of non-ideal gas,
	since the latter requires thermal conductivity

2011-08-23  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* src/hs-collision-integrals.lisp (omega-hs-coeff, omegaLS-hs*):
	Renamed omega-hs-coeff into omegaLS-hs*

2011-08-18  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* src/tc-external-interface.lisp: Added unit tests for O2 Cv and
	Cp.  Pass.

	* src/kt-transport-coeffs.lisp (lambda-0-lj): Added uni test for
	O2 thermal conductivity.  It fails by a factor of 1.27

2011-08-17  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* src/lemmon-jacobsen-2004-coeffs.lisp
	(LJ04-dilute-thermal-conductivity): Returns multiple values of all
	terms in approximation.  Updated doc

	* src/tc-external-interface.lisp (Cv, Cv&Cp): Renamed test Cv into
	Cv&Cp.  Added documentation.  Verified correctness of CRC values.

2011-08-16  Mirko Vukovic  <mirko.vukovic@us.tel.com>

	* README.org: Added more details to warning about thermal
	conductivity calculations.

	* src/environment-setup.lisp (+pi+): +pi+ is single precision
	float.  It's used forced me to rearange the order of calculation
	in lennard-jones mu-0 to prevent floating underflow.  When pi was
	double float, this problem did not arise.

	* src/kt-transport-coeffs.lisp (mu-0): Fixed order of calculations
	to prevent an underflow that resulted in mu-0=0.0.  Fixed both hs
	and lj methods

	* src/transport-coeffs-external-interface.lisp (D-0): Fixed bug in
	funcall to D-0.  Had two arguments instead of required three

	* src/lemmon-jacobsen-2004-coeffs.lisp (Ni): Added error check
	(): Added copyright header
	(): Added documentation

	* src/tc-external-interface.lisp (Cv): For species for which no
	method is defined, I try calculating Cp and then subract N kb (see
	doc).
	(Cv): Added unit tests against CRC tables

	* src/kt-transport-coeffs.lisp (mu-0 lambda-0): Fixed calling
	structure to lj-coeffs and to omega-22 (now called omega-22* to
	better conform to the start notation for reduced collision
	integrals.
	(): Added unit tests for mu-0 and lambda-0 based on lennard-jones
	calculations.  The test for lambda-0 for N2 fails because the

transport-coefficients/mcc-thermal-diffusivity.lisp
;; Mirko Vukovic
;; Time-stamp: <2012-05-27 22:27:50 mcc-thermal-diffusivity.lisp>
;; 
;; Copyright 2011 Mirko Vukovic
;; Distributed under the terms of the GNU General Public License
;; 
;; This program is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.
;; 
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 
;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

(in-package :tc)

(defun at-alpha (Phi-alpha Phi-beta
			m-alpha m-beta
			nu6-alpha-beta nu6-beta-alpha)
  "Sharipov&Kalempa, 2002 (77)

Used in thermal diffusion ratio (73)

I replaced the dependence on nu6-1-2 with nu6-alpha-beta.  That seems
to make more sense"
  (/ (+ Phi-beta (* (sqrt (/ m-alpha m-beta))
		    nu6-alpha-beta))
     (mcc-tc-coeff-denom Phi-alpha Phi-beta
			 nu6-alpha-beta nu6-beta-alpha)))


(define-test at-alpha
  (let ((lisp-unit:*epsilon* 1e-4))
    ;; First batch focuses on PHI-beta in the numerator and
    ;; PHI-alpha/beta in the denumerator by setting all the collision
    ;; frequences to zero.  m-beta =1 to prevent /0 in numerator
    (assert-numerical-equal
     1
     (at-alpha 1 1 0 1 0 0) "simple")
    (assert-numerical-equal
     1
     (at-alpha 1 2 0 1 0 0) "PHI-beta-cancellation/simple")
    (assert-numerical-equal
     1/2
     (at-alpha 2 2 0 1 0 0) "PHI-alpha/simple")
    ;; Second batch adds the mass-fraction terms in the numerator
    (assert-numerical-equal
     2
     (at-alpha 1 1 1 1 1 0) "m-alpha/m-beta * nu6")
    (assert-numerical-equal
     3
     (at-alpha 1 1 4 1 1 0) "m-alpha")
    (assert-numerical-equal
     2
     (at-alpha 1 1 4 4 1 0) "m-beta")
    (assert-numerical-equal
     3
     (at-alpha 1 1 4 4 2 0) "nu6 in numerator")
    ;; Third batch focuses on the nu6 terms in the denominator
    (assert-numerical-equal
     1
     (at-alpha 2 1 0 1 1 1) "nu6 terms")
    (assert-numerical-equal
     -1
     (at-alpha 2 1 0 1 3 1) "nu6-alpha-beta")
    (assert-numerical-equal
     -1
     (at-alpha 2 1 0 1 1 3) "nu6-alpha-beta")))


(defmethod alpha-T-1 ((model (eql 'mcc))
		      (s1 lennard-jones-6/12-potential)
		      (s2 lennard-jones-6/12-potential)
		      p-1 p-2 temp &key)
  (let ((m-1 (* (mass s1) +amu+))
	(m-2 (* (mass  s2) +amu+))
	(n-1 (n p-1 temp))
	(n-2 (n p-2 temp))
	(c-1-1 (make-collision-parameters s1 s1))
	(c-1-2 (make-collision-parameters s1 s2))
	(c-2-1 (make-collision-parameters s2 s1))
	(c-2-2 (make-collision-parameters s2 s2)))
    (with-f-alpha-beta ((1 2 1 2) (1 1 2 2)
			(m* m-alpha m-beta))
      (with-f-alpha-beta ((1 2 1 2) (1 1 2 2)
			  (omega-11 c-alpha-beta temp)
			  (omega-12 c-alpha-beta temp)
			  (omega-13 c-alpha-beta temp)
			  (omega-22 c-alpha-beta temp))
	(with-f-alpha-beta
	    ((1) (2)
	     (nu2 m*-alpha-beta m-alpha n-beta
		  omega-11-alpha-beta
		  omega-12-alpha-beta))
	  (with-f-alpha-beta
	      ((1 2 1 2) (1 1 2 2)
	       (nu5 m*-alpha-beta m-alpha m-beta n-beta
		    omega-11-alpha-beta
		    omega-22-alpha-beta
		    omega-12-alpha-beta
		    omega-13-alpha-beta)
	       (nu6 m*-alpha-beta m-alpha m-beta
		    n-beta omega-11-alpha-beta
		    omega-22-alpha-beta
		    omega-12-alpha-beta
		    omega-13-alpha-beta))
	    (with-f-alpha-beta*
		((1 2) (2 1)
		 (phi nu5-alpha-alpha
		      nu6-alpha-alpha
		      nu5-alpha-beta)
		 (at-alpha phi-alpha-beta phi-beta-alpha
			   m-alpha m-beta
			   nu6-alpha-beta nu6-beta-alpha))
	      (let ((n (+ n-1 n-2)))
		(* -1.25 (/ (* n nu2-1-2)
			    n-2)
		   (- at-alpha-1-2 (* (expt (/ m-1 m-2) 2)
				      at-alpha-2-1)))))))))))


(defmethod alpha-T-1 ((model (eql 'mcc))
		      (s1 hard-sphere-potential)
		      (s2 hard-sphere-potential)
		      p-1 p-2 temp &key)
  (let ((m-1 (* (mass s1) +amu+))
	(m-2 (* (mass  s2) +amu+))
	(n-1 (n p-1 temp))
	(n-2 (n p-2 temp))
	(c-1-1 (make-collision-parameters s1 s1))
	(c-1-2 (make-collision-parameters s1 s2))
	(c-2-1 (make-collision-parameters s2 s1))
	(c-2-2 (make-collision-parameters s2 s2)))
    (with-f-alpha-beta ((1 2 1 2) (1 1 2 2)
			(m* m-alpha m-beta))
      (with-f-alpha-beta ((1 2 1 2) (1 1 2 2)
			  (omega-11 c-alpha-beta temp)
			  (omega-12 c-alpha-beta temp)
			  (omega-13 c-alpha-beta temp)
			  (omega-22 c-alpha-beta temp))
	(with-f-alpha-beta
	    ((1) (2)
	     (nu2 m*-alpha-beta m-alpha n-beta
		  omega-11-alpha-beta
		  omega-12-alpha-beta))
	  (with-f-alpha-beta
	      ((1 2 1 2) (1 1 2 2)
	       (nu5 m*-alpha-beta m-alpha m-beta n-beta
		    omega-11-alpha-beta
		    omega-22-alpha-beta
		    omega-12-alpha-beta
		    omega-13-alpha-beta)
	       (nu6 m*-alpha-beta m-alpha m-beta
		    n-beta omega-11-alpha-beta
		    omega-22-alpha-beta
		    omega-12-alpha-beta
		    omega-13-alpha-beta))
	    (with-f-alpha-beta*
		((1 2) (2 1)
		 (phi nu5-alpha-alpha
		      nu6-alpha-alpha
		      nu5-alpha-beta)
		 (at-alpha phi-alpha-beta phi-beta-alpha
			   m-alpha m-beta
			   nu6-alpha-beta nu6-beta-alpha))
	      (let ((n (+ n-1 n-2)))
		(* -1.25 (/ (* n nu2-1-2)
			    n-2)
		   (- at-alpha-1-2 (* (expt (/ m-1 m-2) 2)
				      at-alpha-2-1)))))))))))

(define-test alphat-tc-vs-gkf/lj
  "Comparison between the code in TRANSPORT-COEFFICIENTS and
  GAS-KINETIC-FORMULARY"
  (let ((n1 0.9e22)
	(n2 0.1e22)
	(s1 (make-species-lennard-jones-6/12-potential :Ar))
	(temp 300d0)
	(lisp-unit:*epsilon* 1e-4))
    (let ((s2 (make-species-lennard-jones-6/12-potential :Ar))
	  (p (p (+ n1 n2) temp))
	  (p1 (p n1 temp))
	  (p2 (p n2 temp)))
      (assert-numerical-equal (gkf::alpha-t s1 s2 n1 n2 temp)
			      (alpha-t-1 'MCC s1 s2 p1 p2 temp)))
    (let ((s2 (make-species-lennard-jones-6/12-potential :He))
	  (p (p (+ n1 n2) temp))
	  (p1 (p n1 temp))
	  (p2 (p n2 temp)))
      (assert-numerical-equal (gkf:alpha-t s1 s2 n1 n2 temp)
			      (alpha-t-1 'MCC s1 s2 p1 p2 temp)))))